abi_stable/src/erased_types/trait_objects.rs

```diff
@@ -45,21 +48,96 @@ impl<'a> HasherObject<'a> {
     pub fn as_mut<'b: 'a>(&'b mut self) -> HasherObject<'b> {
         Self {
             this: self.this.reborrow(),
-            hash_slice: self.hash_slice,
+            write_fns: self.write_fns,
             finish: self.finish,
         }
     }
 }
 
+macro_rules! impl_write {
+    ( $(($ty:ty, $fn:ident)),* ) => {
+        $(
+            fn $fn(&mut self, val: $ty) {
+                unsafe { (self.write_fns.$fn)(self.this.reborrow(), val) }
+            }
+        )*
+    }
+}
+
 impl<'a> Hasher for HasherObject<'a> {
     fn finish(&self) -> u64 {
         unsafe { (self.finish)(self.this.as_rref()) }
     }
     fn write(&mut self, bytes: &[u8]) {
-        unsafe { (self.hash_slice)(self.this.reborrow(), bytes.into()) }
+        unsafe { (self.write_fns.write)(self.this.reborrow(), bytes.into()) }
+    }
+
+    impl_write!(
+        (i16, write_i16),
+        (i32, write_i32),
+        (i64, write_i64),
+        (i8, write_i8),
+        (isize, write_isize),
+        (u16, write_u16),
+        (u32, write_u32),
+        (u64, write_u64),
+        (u8, write_u8),
+        (usize, write_usize)
+    );
+
+    fn write_i128(&mut self, val: i128) {
+        todo!()
```

Comment:
Why `todo!()`, doesn't the default implementation work here(and for `write_u128`)?

Reply:
It's in the other file, but commented out. As far as I know, there's no support for `i128` or `u128`, so these just don't compile when you uncomment it:

```rust
error[E0599]: the associated item `STABLE_ABI` exists for struct `GetTypeLayoutCtor<i128>`, but its trait bounds were not satisfied
   --> abi_stable/src/internal_macros.rs:6:37
    |
6   |         __GetTypeLayoutCtor::<$ty>::STABLE_ABI
    |                                     ^^^^^^^^^^ associated item cannot be called on `GetTypeLayoutCtor<i128>` due to unsatisfied trait bounds
    |
   ::: abi_stable/src/abi_stability/stable_abi_trait.rs:136:1
    |
136 | pub struct GetTypeLayoutCtor<T>(T);
    | ----------------------------------- associated item `STABLE_ABI` not found for this
    |
   ::: abi_stable/src/erased_types/trait_objects.rs:104:10
    |
104 | #[derive(StableAbi)]
    |          --------- in this derive macro expansion
    |
    = note: the following trait bounds were not satisfied:
            `i128: stable_abi_trait::StableAbi`
    = note: this error originates in the macro `$crate::_sabi_type_layouts` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the associated item `STABLE_ABI` exists for struct `GetTypeLayoutCtor<u128>`, but its trait bounds were not satisfied
   --> abi_stable/src/internal_macros.rs:6:37
    |
6   |         __GetTypeLayoutCtor::<$ty>::STABLE_ABI
    |                                     ^^^^^^^^^^ associated item cannot be called on `GetTypeLayoutCtor<u128>` due to unsatisfied trait bounds
    |
   ::: abi_stable/src/abi_stability/stable_abi_trait.rs:136:1
    |
136 | pub struct GetTypeLayoutCtor<T>(T);
    | ----------------------------------- associated item `STABLE_ABI` not found for this
    |
   ::: abi_stable/src/erased_types/trait_objects.rs:104:10
    |
104 | #[derive(StableAbi)]
    |          --------- in this derive macro expansion
    |
    = note: the following trait bounds were not satisfied:
            `u128: stable_abi_trait::StableAbi`
    = note: this error originates in the macro `$crate::_sabi_type_layouts` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0599`.
warning: `abi_stable` (lib) generated 2 warnings
error: could not compile `abi_stable` due to 2 previous errors; 2 warnings emitted
```

I did this a while ago so I might have forgotten what was fully going on, though. But it seems like these types don't implement `StableAbi`, so they can't be in a function in `WriteFns`.